pkg/kudoctl/cmd/upgrade.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1alpha1"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/cmd/install"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/util/kudo"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/util/repo"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var (
+	upgradeExample = `
+		The upgrade argument must be a name of the package in the repository, a path to package in *.tar.gz format,
+		or a path to an unpacked package directory.
+
+		# Upgrade flink instance dev-flink to the latest version
+		kubectl kudo upgrade flink --instance dev-flink
+		
+		*Note*: should you have a local "flink" folder in the current directory it will take precedence over the remote repository.
+
+		# Upgrade flink to the version 1.1.1
+		kubectl kudo upgrade flink --instance dev-flink --version 1.1.1
+
+		# By default parameters are all reused from the previous installation, if you need to modify, use -p
+		kubectl kudo upgrade flink --instance dev-flink -p param=xxx`
+)
+
+type options struct {
+	InstanceName   string
+	Namespace      string
+	PackageVersion string
+	Parameters     map[string]string
+}
+
+// defaultOptions initializes the install command options to its defaults
+var defaultOptions = &options{
+	Namespace: "default",
+}
+
+// newUpgradeCmd creates the install command for the CLI
+func newUpgradeCmd() *cobra.Command {
+	options := defaultOptions
+	var parameters []string
+	upgradeCmd := &cobra.Command{
+		Use:     "upgrade <name>",
+		Short:   "Upgrade KUDO package.",
+		Long:    `Upgrade KUDO package from current version to new version.`,
+		Example: upgradeExample,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// Prior to command execution we parse and validate passed parameters
+			var err error
+			options.Parameters, err = install.GetParameterMap(parameters)
+			if err != nil {
+				return errors.WithMessage(err, "could not parse parameters")
+			}
+			return runUpgrade(args, options)
+		},
+		SilenceUsage: true,
+	}
+
+	upgradeCmd.Flags().StringVar(&options.InstanceName, "instance", "", "The instance name.")
+	upgradeCmd.Flags().StringArrayVarP(&parameters, "parameter", "p", nil, "The parameter name and value separated by '='")
+	upgradeCmd.Flags().StringVar(&options.Namespace, "namespace", "default", "The namespace where the instance you want to upgrade is installed in. (default \"default\"")
+	upgradeCmd.Flags().StringVarP(&options.PackageVersion, "version", "v", "", "A specific package version on the official repository. When installing from other sources than official repository, version from inside operator.yaml will be used. (default to the most recent)")
+
+	const usageFmt = "Usage:\n  %s\n\nFlags:\n%s"
+	upgradeCmd.SetUsageFunc(func(cmd *cobra.Command) error {
+		fmt.Fprintf(upgradeCmd.OutOrStderr(), usageFmt, upgradeCmd.UseLine(), upgradeCmd.Flags().FlagUsages())
+		return nil
+	})
+	return upgradeCmd
+}
+
+func validate(args []string, options *options) error {
```

Comment:
this feels more like a validateCmd or validateArgs...   find a validate in upgrade I was thinking validation rules around upgrade which this isn't.   much of that is in `upgrade` where version comparisons are done.

simply mentioning... it is a a weak comment on my part... it looks fine just what was in my head as I reviewed.